### ISSUE TITLE
Fail to deploy k8s cluster 1.8 on vSphere

### ISSUE THREAD

**Issue by abrarshivani** [kubernetes/kubernetes-anywhere#495]
Logs:-
```
kubeanywhere- null_resource.kubernetes-node3 (remote-exec): Created symlink from /etc/systemd/system/multi-user.target.wants/kubelet.service to /etc/systemd/system/kubelet.service.
null_resource.kubernetes-node3: Creation complete (ID: 2083685943369692449)
Error applying plan:

1 error(s) occurred:

* null_resource.kubernetes-master: 1 error(s) occurred:

* Script exited with non-zero exit status: 1
```

**Comment by tmbull** [comment]
The documentation states that 1.6.5 is recommended, so I am not surprised that this does not work, but 1.6.5 was released around 6 months ago now. So, I am also curious as to what the roadmap is for supporting newer releases.

**Comment by tmbull** [comment]
FWIW, the above error was caused in my case by an invalid version tag (`1.8.2` vs `v1.8.2`). 

After setting the version correctly, the deployment seems to work until validation which eventually times out: 

```
State path:
Kubectl binary with version - v1.8.2 doesn't exist. So downloading it.
Kubectl binary is not available at https://storage.googleapis.com/kubernetes-release/release/v1.8.2/bin/linux/amd64/kubectl
make[1]: Leaving directory '/opt/kubernetes-anywhere'
KUBECONFIG="$(pwd)/phase1/vsphere/kubernetes/kubeconfig.json" ./util/validate
Validation: Expected 6 (workers + master) healthy nodes; found 0. (10s elapsed)
Validation: Expected 6 (workers + master) healthy nodes; found 0. (20s elapsed)
Validation: Expected 6 (workers + master) healthy nodes; found 0. (30s elapsed)
...
Validation: Expected 6 (workers + master) healthy nodes; found 0. (1190s elapsed)
Validation: Expected 6 (workers + master) healthy nodes; found 0. (1200s elapsed)
Timeout exceeded.
Makefile:61: recipe for target 'validate' failed
```

**Comment by serkilov** [comment]
Same for 1.8.4

```
State path:
Kubectl binary with version - v1.8.4 doesn't exist. So downloading it.
--2017-12-07 10:38:56--  https://storage.googleapis.com/kubernetes-release/release/v1.8.4/bin/linux/amd64/kubectl
Resolving storage.googleapis.com... 172.217.10.16, 2607:f8b0:4006:802::2010
Connecting to storage.googleapis.com|172.217.10.16|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 52334123 (50M) [application/octet-stream]
Saving to: '/usr/local/bin/kubectl_v1.8.4'

/usr/local/bin/kubectl_v1.8.4                               100%[========================================================================================================================================>]  49.91M  54.2MB/s    in 0.9s

2017-12-07 10:38:57 (54.2 MB/s) - '/usr/local/bin/kubectl_v1.8.4' saved [52334123/52334123]

make[1]: Leaving directory '/opt/kubernetes-anywhere'
...
Validation: Expected 2 (workers + master) healthy nodes; found 0. (1200s elapsed)
Timeout exceeded.
Makefile:61: recipe for target 'validate' failed
make: *** [validate] Error 1
```

**Comment by TimMeade** [comment]
any update on this?  Just did v1.8.5 and same thing.   times out on validation....

**Comment by TimMeade** [comment]
I have tried to deploy v1.8.5, v1.7.11, and v1.6.5 and all have failed similarly.

it appears to be with the connection between kubectl and the cluster.   breaking down what is in the script i ran:

```
kubectl --kubeconfig phase1/vsphere/k8s-prod-atl/kubeconfig.json get nodes
```
and recieved
```
The connection to the server 10.2.102.193 was refused - did you specify the right host or port?
```
I verified the ip is the same as what is in the kubeconfig.json file and is the ip for the master node.  I can ping the master node from the setup docker machine.  

Any ideas?

**Comment by D3nn** [comment]
I don't know if it's the same issue, but I got this advice from @divyenpatel on another issue and it helped:

For 1.8 kubernetes releases you can use this new ignition image cnastorage/k8s-ignition:v1.8-dev-release This ignition image uses new kubelet flags for 1.8 releases

Also, you might want to ssh to the master node and run `systemctl status -l kubelet` to see if there are any other issues.  I had vsphere cluster hosts whose times weren't in sync and this caused issues with the certificates not yet being valid.

**Comment by TimMeade** [comment]
Thanks @D3nn ... 

Good pick on time.  One server was off by 3 minutes and it's ntp server was blank. Odd, but fixed.  All now synced.

Tried without changing the ignition image.   All the VM's are using tons of CPU, but no changes.  Running the systemctl check and it just times out

So I dropped in the new ignition VM and now they are using much less CPU, but when I run the systemctl check on the master:

```
kubelet.service
     Loaded:  no-found (Reason:  no such file or directory)
     Active:  inactive (dead)
```

So progress it seems, but no solution.

**Comment by TimMeade** [comment]
Well... dang.

Blew EVERYTHING away.   restarted docker image for kube anywhere.  
Redid make config getting everything correct and using docker.io/cnastorage/k8s-ignition:v1.8-dev-release for the ignition image.  Verified time on hosts are in sync.

Created the cluster in under 3 minutes.   Appears everything working with v1.8.5   starting testing.

**Comment by D3nn** [comment]
Glad to hear it @TimMeade.  Your experience pretty much mirrored mine, though I was able to:
* make destroy
* edited .config using a text editor (with the correct values)
* make clean
* make config .config (and just held down the enter key until all my new defaults were chosen)
* make deploy

Then all was happy and the cluster was up pretty quickly.

**Comment by tmbull** [comment]
@D3nn Thanks for the tip on the ignition image. I updated my .config accordingly and was able to deploy 1.8.5 successfully.

**Comment by ipnextgen** [comment]
Hi,

Mind doing a step-by-step to deploy to 1.8? Meanwhile the project gets updated

Thanks!

**Comment by D3nn** [comment]
@ipnextgen I can't do much better than what's already documented [here](https://github.com/kubernetes/kubernetes-anywhere/blob/master/phase1/vsphere/README.md), but I can offer some crib notes:
1. Make sure the vSphere ESX hosts' times are in sync (using NTP).  Having out of sync times will make the Kubernetes-Anywhere install fail.
1.  If you don't already know, figure out the names of the datacenter, cluster, and resource pool.  If you're not sure, in the vsphere inventory "Hosts and Clusters" view, click on the various items and you'll see in the right-hand panel a Summary page which has headings like "What is a DataCenter" (or Cluster or Resource Pool).  I say this mainly because our vSphere environment was setup by our IT department so it wasn't immediately obvious to me when I began this process.
1.  Also note when you have a Folder that houses one of the above objects.  Ours has a top level folder (named "Colorado Springs") and a Datacenter confusingly named "Resource Pools."  So our Datacenter name was "Colorado Spring/Resource Pools", which wasn't exactly intuitive to me.
1.  When creating the OVA, click on your Resource Pool object, then click File in the top menu and click "Deploy OVF Template".    Then cut-n-paste from the OVA template link in the README appropriate to your cluster into the "Deploy from a file or URL" input box (and then follow the prompts).  This is where Kubernetes-Anywhere expects to find it.
1. When running the "make config", enter the Datacenter, Cluster and Resource Pool when asked (most of the other stuff should be decent defaults, save for a few things used in the bootstrap phase)
1. For the "installer container" question, use 'docker.io/cnastorage/k8s-ignition:v1.8-dev-release' 
1. I didn't have a gcr account, so for the "docker registry" question, I used 'docker.io/googlecontainers' (there's no '-' between 'google' and 'containers' on the docker hub hosted organization).
1. For the version, I chose the latest 1.8 version available (which was 1.8.5).

When you're done with the config, run `make deploy`.  If things go wrong, see where it failed and make adjustments as necessary using the procedure I noted three comments up from this.

Good luck!

**Comment by TimMeade** [comment]
I just wanted to append to this.  This still works using v1.9.0 for the K8s version.....

**Comment by marrik96** [comment]
Thanks for the info @TimMeade.  I have k8s cluster v1.6.5 running on vSphere 6.0 fine but was interested in trying out a newer version.  I also wish there was a simple way to upgrade the cluster.  I guess at this time it's just as easy to destroy and create from scratch.

**Comment by vdavy** [comment]
Got it worked with :
* kubernetes 1.9.1 
* this param : `.phase2.installer_container="docker.io/cnastorage/k8s-ignition:v1.8-dev-release"`

Enjoy :smile:

**Comment by bbratchiv** [comment]
@vdavy , thanks, that worked, but master node has the Status: Ready,SchedulingDisabled
Any idea how make is scheduled? It's not about taints, this was done during deployment

**Comment by vdavy** [comment]
Sorry @bbratchiv - absolutly not idea

**Comment by divyenpatel** [comment]
@bbratchiv you need to uncordon master node. See below.

```
$ kubectl uncordon kubernetes-master
node "kubernetes-master" uncordoned
```

You can enable scheduling permanently on the master node by setting `--register-schedulable=true` in `/etc/systemd/system/kubelet.service`. Currently this flag is set to `false`.

**Comment by bbratchiv** [comment]
@divyenpatel , wow! Thank you!

**Comment by joerawr** [comment]
Thanks for everyone's help on this.  I just ran this successfully with v1.8.8.  (I previously had v1.6.5 up but was running into some issues, so decided to trash the whole thing and start again).

Some possibly helpful notes:
`docker pull divyen/kubernetes-anywhere:11302017`
`docker run -it -v /tmp:/tmp --rm --env="PS1=[container]:\w> " --net=host divyen/kubernetes-anywhere:11302017 /bin/bash`

@D3nn 
For docker registry, I tried "docker.io/googlecontainers" and got an error that it couldn't pull hyperkube.  changed it back to "gcr.io/google-containers" and re-ran.  I don't have any gcr credentials or anything configured for that.  It just worked.  Noting it here in case it helps someone else.

```null_resource.k8sc1-node4 (remote-exec): Pulling repository docker.io/googlecontainers/hyperkube-amd64
null_resource.k8sc1-node2 (remote-exec): Pulling repository docker.io/googlecontainers/hyperkube-amd64
null_resource.k8sc1-node4 (remote-exec): Error: image googlecontainers/hyperkube-amd64 not found
null_resource.k8sc1-node4 (remote-exec): Failed to docker pull hyperkube
```

There were a few new things not in the current guide, leaving them both default seems to work fine:
Under:
```
*
* Phase 2: Node Bootstrapping
*
```
I left kube-proxy default
`kube-proxy mode: iptables or ipvs (phase2.proxy_mode) [iptables] (NEW) `

And under:
```
* Phase 3: Deploying Addons
*
```
I left CNI plugin blank:
   ` CNI plugin to use (phase3.cni) [] (NEW) `

Remember to ```make clean``` before you re-run the make deploy.
I forgot to do that after making an edit to the .config and it spun up the 5 nodes, but could only talk to the master, so I was stuck at 
```Validation: Expected 5 (workers + master) nodes; found 1. (40s elapsed) ...```  trying to figure out what was wrong...  _make clean_

**Comment by WuTangDangie** [comment]
I was unable to deploy a version newer than v1.6.5, but this issue board answered a lot of my questions.  Thanks for that.  

My remaining question is, what do I do with old cluster versions?   If I have workloads running on older versions, how do I update the cluster?  And if the answer is, "Don't update, create new.", then how do I move from the old to the new?  

I'm just beginning to dive into this.  Any help would be great.  Thanks!

**Comment by tickers** [comment]
Just wanted to say thanks!! Using the added info from here I was able to get to 1.9.3 with VMWare 6.5. Volumes are working!! 
Here's my config
To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
[container]:/opt/kubernetes-anywhere> cat .config
#
# Automatically generated file; DO NOT EDIT.
# Kubernetes Minimal Turnup Configuration
#

#
# Phase 1: Cluster Resource Provisioning
#
.phase1.num_nodes=4
.phase1.cluster_name="kubernetes"
.phase1.ssh_user=""
.phase1.cloud_provider="vsphere"

#
# vSphere configuration
#
.phase1.vSphere.url="192.168.0.101"
.phase1.vSphere.port=443
.phase1.vSphere.username="administrator@fakeuser"
.phase1.vSphere.password="SomeShadyFakestuff"
.phase1.vSphere.insecure=y
.phase1.vSphere.datacenter="mydatacenter"
.phase1.vSphere.datastore="jbod" <-- I use an Old HP Microserver with passthru for storage :-)
.phase1.vSphere.placement="cluster"
.phase1.vSphere.cluster="clusterofmine"
.phase1.vSphere.useresourcepool="yes"
.phase1.vSphere.resourcepool="kubernetes"
.phase1.vSphere.vmfolderpath="kubernetes"
.phase1.vSphere.vcpu=8
.phase1.vSphere.memory=4096
.phase1.vSphere.network="Openstack"
.phase1.vSphere.template="KubernetesAnywhereTemplatePhotonOS.ova"
.phase1.vSphere.flannel_net="172.1.0.0/16"

#
# Phase 2: Node Bootstrapping
#
.phase2.kubernetes_version="v1.9.3"
.phase2.provider="ignition"
.phase2.installer_container="docker.io/cnastorage/k8s-ignition:v1.8-dev-release-with-kubelet-fix"
.phase2.docker_registry="gcr.io/google-containers"

#
# Phase 3: Deploying Addons
#
.phase3.run_addons=y
.phase3.kube_proxy=y
.phase3.dashboard=y
.phase3.heapster=y
.phase3.kube_dns=y
# .phase3.weave_net is not set

After I had to install kubectl on the Master using - https://coreos.com/tectonic/docs/latest/tutorials/kubernetes/configure-kubectl.html

Used the fix provided by divyenpatel commented on Jan 24

Tested the volume creation using - https://github.com/kubernetes/examples/blob/master/staging/volumes/vsphere/README.md#storage-class

Hope that helps someone else out!

**Comment by tickers** [comment]
One thin I have found and I am trying to find the root cause is this. 
I can run through testing, working on my CKA, so Kubernetes gets hammered pretty good with ZERO issues. But after letting it sit over night do nothing I get this when I go to the web GUI
443/api/v1/nodes: dial tcp 10.0.0.1:443: getsockopt: connection refused when I select nodes pods or what ever. The GUI is up just it's contents are having issues. Make sense?
Any thoughts on what I can check and how?

**Comment by WuTangDangie** [comment]
Any luck figuring out your connection issue @tickers ?  I'm seeing the same thing in my environment: Fine when frequently accessed, but unreachable after inactivity (usually overnight). 

Thanks.

**Comment by tickers** [comment]
WuTangDangie - Nope still trying to find the cause.

**Comment by tickers** [comment]
Moved onto v1.10.0
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:44:10Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

# Phase 2: Node Bootstrapping
#
.phase2.kubernetes_version="v1.10.0"
.phase2.provider="ignition"
.phase2.installer_container="docker.io/cnastorage/k8s-ignition:v1.8-dev-release-with-kubelet-fix"
.phase2.docker_registry="gcr.io/google-containers"

Everything is working though I would like to upgrade the dashboard and the client at some point and also test to see if this version locks up over night.

**Comment by tickers** [comment]
v1.10.0 has been running steady for almost two days with mysql, wordpress and 4 persistent storage running. Also upgraded the dashboard to 1.8.3 though I stayed with the plain http vs the https install.  So far it's hanging tough so next I will toss in HELM and run a few installs that way and push it more towards the edge and see if any falls off.  Each node has 8vcpu and 4096 of mem fyi.

**Comment by marrik96** [comment]
That's great to hear tickers.  I will be upgrading soon.

**Comment by tickers** [comment]
seems was hit with the error
error_outline Internal Server Error (500)
Get https://10.0.0.1:443/api/v1/replicationcontrollers: dial tcp 10.0.0.1:443: getsockopt: connection refused
Resolved by rebooting the entire cluster via vmware. I also removed a Helm app I had deployed 
that was stuck.  Getting helm working was a chore on Photon 1, had to instal a new socat and openssl for a libssl.so.1.1 missing error. Not sure but it may be the openssl 1.1 and Photon 1 have issues..duno.

If your brave
Helm added steps
install compiler and tools — https://github.com/vmware/photon/issues/693 
tdnf install lvm2 gawk make libgcc-devel libstdc++-devel util-linux linux-api-headers autoconf libgcc-devel linux-api-headers bison openssl-devel automake acl libacl-devel zlib-devel wget python2-devel libxml2-devel diffutils gcc glibc-devel binutils tar glibc-tools sqlite-devel userspace-rcu-devel python-xml wget git tar

Install socat — https://www.cyberciti.biz/faq/linux-unix-tcp-port-forwarding/ 
wget http://www.dest-unreach.org/socat/download/socat-1.7.3.2.tar.gz
 tar -zxvf socat-1.7.3.2.tar.gz
cd socat-1.7.3.2
./configure
 make
 make install

https://github.com/openssl/openssl/blob/master/INSTALL
https://www.openssl.org/source/
wget https://www.openssl.org/source/openssl-1.1.0h.tar.gz
tar -zxvf openssl-1.1.0h.tar.gz
./configure
make
make test
make install

If you have missing libs check you 
  cat /etc/ld.so.conf
  vi /etc/ld.so.conf add missing dir's as need
  then run
  ldconfig

May upgrade to photon 2 if it does it again in case there some compatibility issue I am missing. 
All of my nodes and master are snapshots so I can roll forward or back as needed for testing.

Cheers

**Comment by aaron-comyn** [comment]
Tried the scripts and @tickers settings on kubernetes v1.11.0: installation failed stating "docker.io/cnastorage/k8s-ignition:v1.8-dev-release-with-kubelet-fix" is not a valid repository/tag.

Pulling the image works fine from the nodes, so perhaps something with the remote execution is misinterpreting the name... Reverting to kubernetes v1.10.0 avoided the issue :)

**Comment by armontero** [comment]
I just wanted to add to this thread and say that I was able to get 1.11.1 installed by using the .phase2.installer_container="docker.io/cnastorage/k8s-ignition:v1.8-dev-release" recommendation I found this thread.

I also just wanted to say how awesome this project is and that I hope the devs continue to support this project going forward.  Being able to stand up any sized cluster in minutes is just awesome.  Thank you for putting this together!

**Comment by alexometis** [comment]
Another addition - v1.6.5 worked fine for me initially but I needed newer API versions for my application.
I tried 1.9.10 and 1.8.15 but neither worked initially - then I came across this issue. 
I then tried 1.9.10 and 1.8.15 with the .phase2.installer_container="docker.io/cnastorage/k8s-ignition:v1.8-dev-release" settiing and both failed. 
Finally I tried 1.8.5 as someone mentioned that specific version here and it worked for me.

**Comment by alexometis** [comment]
Another addition - v1.6.5 worked fine for me initially but I needed newer API versions for my application.
I tried 1.9.10 and 1.8.15 but neither worked initially - then I came across this issue. 
I then tried 1.9.10 and 1.8.15 with the .phase2.installer_container="docker.io/cnastorage/k8s-ignition:v1.8-dev-release" settiing and both failed. 
Finally I tried 1.8.5 as someone mentioned that specific version here and it worked for me.
